irohad/ametsuchi/impl/postgres_query_executor.cpp

```diff
@@ -0,0 +1,800 @@
+        OR (SELECT * FROM has_all_perm)
+        OR ('%7%' = '%8%' AND (SELECT * FROM has_domain_perm)) AS perm
+    )");
+
+    return (cmd % bits % creator % perm_str % all_perm_str % domain_perm_str
+            % target_account % getDomainFromName(creator)
+            % getDomainFromName(target_account))
+        .str();
+  }
+
+  void callback(
+      std::vector<std::shared_ptr<shared_model::interface::Transaction>>
+          &blocks,
+      uint64_t block_id,
+      ametsuchi::KeyValueStorage &block_store,
+      std::vector<uint64_t> &result) {
+    auto block = block_store.get(block_id) | [](const auto &bytes) {
+      return shared_model::converters::protobuf::jsonToModel<
+          shared_model::proto::Block>(bytesToString(bytes));
+    };
+    if (not block) {
+      return;
+    }
+
+    std::transform(
+        result.begin(),
+        result.end(),
+        std::back_inserter(blocks),
+        [&](const auto &x) {
+          return std::shared_ptr<shared_model::interface::Transaction>(
+              clone(block->transactions()[x]));
+        });
+  }
+}  // namespace
+
+namespace iroha {
+  namespace ametsuchi {
+    PostgresQueryExecutor::PostgresQueryExecutor(
+        std::unique_ptr<soci::session> sql,
+        std::shared_ptr<shared_model::interface::CommonObjectsFactory> factory,
+        KeyValueStorage &block_store)
+        : sql_(std::move(sql)),
+          block_store_(block_store),
+          factory_(factory),
+          visitor_(*sql_, factory_, block_store_) {}
+
+    QueryExecutorResult PostgresQueryExecutor::validateAndExecute(
+        const shared_model::interface::Query &query) {
+      visitor_.setCreatorId(query.creatorAccountId());
+      auto result = boost::apply_visitor(visitor_, query.get());
+      return clone(result.queryHash(query.hash()).build());
+    }
+
+    bool PostgresQueryExecutor::validate(
+        const shared_model::interface::BlocksQuery &query) {
+      boost::format cmd(R"(%s)");
+      soci::statement st =
+          (sql_->prepare
+           << (cmd % checkAccountRolePermission(Role::kGetBlocks)).str());
+      int has_perm;
+      st.exchange(soci::use(query.creatorAccountId(), "role_account_id"));
+      st.exchange(soci::into(has_perm));
+      st.define_and_bind();
+      st.execute(true);
+
+      return has_perm > 0;
+    }
+
+    PostgresQueryExecutorVisitor::PostgresQueryExecutorVisitor(
+        soci::session &sql,
+        std::shared_ptr<shared_model::interface::CommonObjectsFactory> factory,
+        KeyValueStorage &block_store)
+        : sql_(sql), block_store_(block_store), factory_(factory) {}
+
+    void PostgresQueryExecutorVisitor::setCreatorId(
+        const shared_model::interface::types::AccountIdType &creator_id) {
+      creator_id_ = creator_id;
+    }
+
+    QueryResponseBuilderDone PostgresQueryExecutorVisitor::operator()(
+        const shared_model::interface::GetAccount &q) {
+      using T = boost::tuple<
```

Comment:
Consider using such shortcut to improve readability:
```C++
template <typename... Value>
using QueryReturnType = boost::tuple<boost::optional<Value>..., int>;

using T = QueryReturnType<shared_model::interface::types::AccountIdType,
                          shared_model::interface::types::DomainIdType,
                          shared_model::interface::types::QuorumType,
                          shared_model::interface::types::DetailType,
                          shared_model::interface::types::RoleIdType>;
```

Same for methods below.